File: CS50x/sentimental-credit/credit.py
# import get_string from cs50
from cs50 import get_string
import sys


# calculate sum of digits
def getSum(n):
    sum = 0
    for digit in str(n):
        sum += int(digit)
    return sum


# ask input number from user and check if correct format entered
def cardcheck():
    while True:
        try:
            n = get_string("Number: ")
        except ValueError:
            print("Incorrect input.")
        else:
            if n != "" and len(n) >= 13 and len(n) <= 16:
                return n
            else:
                print("INVALID")
                sys.exit(1)


# calculate sum of double of digits starting from second from the back
def countDouble(n):
    total = 0
    length = len(n)

    while length > 1:
        tempn = int(n[length-2:length-1]) * 2
        total += getSum(tempn)
        length -= 2

    return total


# calculate sum of remaining digits
def countRemainder(n):
    total = 0
    length = len(n)

    while length >= 1:
        if length == 1:
            total += int(n[length-1:length])
            length = 0
        else:
            tempn = int(n[length-1:length])
            total += getSum(tempn)
            length -= 2

    return total


# main
def main():
    number = cardcheck()

    first = countDouble(number)
    second = countRemainder(number)
    total = first + second

    # check if remainder is 0 if not, card number is invalid
    if total % 10 != 0:
        print("INVALID")
        sys.exit(1)

    # case visa
    if len(number) == 13:
        if number[0:1] == "4":
            print("VISA")
            sys.exit(0)
    # case amex
    if len(number) == 15:
        if number[0:2] == "34" or number[0:2] == "37":
            print("AMEX")
            sys.exit(0)
    # check mc or visa
    if len(number) == 16:
        if number[0:2] == "51" or number[0:2] == "52" or number[0:2] == "53" or number[0:2] == "54" or number[0:2] == "55":
            print("MASTERCARD")
            sys.exit(0)
        elif number[0:1] == "4":
            print("VISA")
            sys.exit(0)

    print("INVALID")
    sys.exit(0)


# call main function
main()

File: CS50x/credit/credit.c
#include <cs50.h>
#include <stdio.h>

// constants
const string INVALID = "INVALID\n";

// declare functions
long get_number(void);

int main(void)
{
    long card_number = get_number();
}

// function to get card number and calculate validity and card type
long get_number(void)
{

    long card_number;
    // ask user to input card number
    do
    {
        card_number = get_long("Enter your card number: ");
    }

    while (card_number < 0);

    int length = 0;
    long card_number2 = card_number;
    while (card_number2 > 0)
    {
        card_number2 = card_number2 / 10;
        length++;
    }
    if (length > 16 || length < 13)
    {
        printf("%s", INVALID);
        return 0;
    }
    else
    {
        int digits = 0;
        int calculate_digit = 0;
        int last_digit = 0;
        card_number2 = card_number;
        while (card_number2 > 0)
        {
            last_digit += card_number2 % 10;
            card_number2 = card_number2 / 10;
            int temp_digit = (card_number2 % 10) * 2;
            if (temp_digit > 9)
            {
                printf("Temp digit: %i\n", temp_digit);
                int temp_digit2 = temp_digit % 10;
                printf("Temp digit 2: %i\n", temp_digit2);
                temp_digit = (temp_digit / 10) + temp_digit2;
                printf("Temp digit: %i\n", temp_digit);
            }

            calculate_digit += temp_digit;
            card_number2 = card_number2 / 10;
            digits++;

            printf("Last Digit: %i\n", last_digit);
            printf("Calculate Digit: %i\n", calculate_digit);
        }

        printf("Digit calculation: %i\n", calculate_digit);

        calculate_digit += last_digit;

        /*
        Determine card type

        Card Type       Digits      Starting numbers
        AMEX            15          34, 37
        MASTERCARD      16          51, 52, 53, 54, 55
        VISA            13/16       4
        */

        if (calculate_digit % 10 == 0)
        {
            printf("Number validated: %i\n", calculate_digit);
            printf("Card Number: %li\n", card_number);
            while (card_number >= 100)
            {
                card_number = card_number / 10;
            }
            if (card_number / 10 == 4)
            {
                printf("VISA\n");
            }
            else if (card_number / 10 == 3 && (card_number % 10 == 4 || card_number % 10 == 7))
            {
                printf("AMEX\n");
            }
            else if (card_number / 10 == 5 &&
                     (card_number % 10 == 1 || card_number % 10 == 2 || card_number % 10 == 3 || card_number % 10 == 4 || card_number % 10 == 5))
            {
                printf("MASTERCARD\n");
            }
            else
            {
                printf("%s", INVALID);
            }
        }
        else
        {
            printf("%s", INVALID);
        }
    }

    return card_number;
}